[DO NOT MERGE, BUT DO REVIEW!] HF x PyTorch DDP Blog (#545)
huggingface/blog: pytorch-ddp-accelerate-transformers.md

test_loader = torch.utils.data.DataLoader(test_dset, shuffle=False, batch_size=64)
```

Move the model to the CUDA device:

```python
model = BasicNet().to(device)
```

Build a PyTorch optimizer:

```python
optimizer = optim.AdamW(model.parameters(), lr=1e-3)
```

Before finally creating a simplistic training and evaluation loop that performs one full iteration over the dataset and calculates the test accuracy:

```python
model.train()
for batch_idx, (data, target) in enumerate(train_loader):
    data, target = data.to(device), target.to(device)
    output = model(data)
    loss = F.nll_loss(output, target)
    loss.backward()
    optimizer.step()
    optimizer.zero_grad()
model.eval()
correct = 0
with torch.no_grad():
    for data, target in test_loader:
        data, target = data.to(device), target.to(device)
        output = model(data)
        pred = output.argmax(dim=1, keepdim=True)
        correct += pred.eq(target.view_as(pred)).sum().item()
print(f'Accuracy: {100. * correct / len(test_loader.dataset)}')
```

Typically from here, one could either throw all of this into a python script or run it on a Jupyter Notebook.

> However, how would you then get this to run on say two GPUs if these resources available? Just doing `python myscript.py` will only ever run the script using a single GPU. This is where `torch.distributed` comes into play

Author: Met in the middle on this suggestion and reworded it some

To convert our above code to work within a distributed setup, a few setup configurations must first be defined, detailed in the [Getting Started with DDP Tutorial](https://pytorch.org/tutorials/intermediate/ddp_tutorial.html)

First a `setup` and a `cleanup` function must be declared. This will open up a processing group that all of the compute processes can communicate through

> Note: for this section of the tutorial it should be assumed these are sent in python script files. Later on a launcher using Accelerate will be discussed that removes this necessity

```python
import os
import torch.distributed as dist

def setup(rank, world_size):
    "Sets up the process group and configuration for PyTorch Distributed Data Parallelism"
    os.environ["MASTER_ADDR"] = 'localhost'
    os.environ["MASTER_PORT"] = "12355"

    # Initialize the process group
    dist.init_process_group("gloo", rank=rank, world_size=world_size)

def cleanup():
    "Cleans up the distributed environment"
    dist.destroy_process_group()
```

The last piece of the puzzle is *how do I send my data and model to another GPU?*

> This is where the `DistributedDataParallel` module comes into play. It will copy your model into each GPU, and when `loss.backward()` is called the gradients across all these copies of the model will be averaged, backprop is performed, and each of the weights get updated to have the same weights across all devices.

Reviewer: ```suggestion
This is where the `DistributedDataParallel` module comes into play. It will copy your model onto each GPU, and when `loss.backward()` is called the backpropagation is performed and the resulting gradients across all these copies of the model will be averaged/reduced. This ensures each device has the same weights post the optimizer step.
```

Below is an example of our training setup, refactored as a function, with this capability:

> > Note: Here rank is the overall rank of the GPU based on all other devices, meaning they have a rank of `0 -> n`

Author: Rewording more accurately:

> Note: Here rank is the overall rank of the current GPU compared to all the other GPUs available, meaning they have a rank of `0 -> n-1`

```python
from torch.nn.parallel import DistributedDataParallel as DDP

def train(model, rank, world_size):
    setup(rank, world_size)
    model = model.to(rank)
    ddp_model = DDP(model, device_ids=[rank])
    optimizer = optim.AdamW(ddp_model.parameters(), lr=1e-3)
    # Train for one epoch
    model.train()
    for batch_idx, (data, target) in enumerate(train_loader):
        data, target = data.to(device), target.to(device)
        output = model(data)
        loss = F.nll_loss(output, target)
        loss.backward()
        optimizer.step()
        optimizer.zero_grad()
    cleanup()

> ```

Reviewer: I believe changes that are necessary for dataloaders are left out. That is a significant detail and quite a bit of a hassle.  Accelerate abstracts and that is great for users. So, good to detail this too. For normal datasets, that would mean using the `DistributedSampler` sampler https://pytorch.org/docs/stable/data.html#torch.utils.data.distributed.DistributedSampler.

> 

Reviewer: Also, details on handling outputs during evaluation, i.e, gathering from all the processes is also quite a bit of hassle and an important detail when woking in distributed setting else evaluations won't be reliable hampering the experimentation process. Accelerate abstracts this too making life simple for users.
Without Accelerate one would have code like below:
```
gather_predictions = [predictions.clone() for _ in range(dist.get_world_size())]
torch.distributed.all_gather(gather_predictions, predictions)
batch_predictions =  torch.cat(gather_predictions, dim=0)

gather_labels = [batch["labels"].clone() for _ in range(dist.get_world_size())]
torch.distributed.all_gather(gather_labels, batch["labels"])
batch_labels =  torch.cat(gather_labels, dim=0)

metric.add_batch(
    predictions=batch_predictions.detach().cpu().numpy(),
    references=batch_labels.detach().cpu().numpy(),
)

```

The optimizer needs to be declared based on the model *on the specific device* for all of the gradients to properly be calculated.

Lastly, to run the script PyTorch has a convient `torchrun` command line module that can help. Just pass in the number of nodes it should use as well as the script to run and you are set:

```bash
torchrun --nproc_per_nodes=2 --nnodes=1 example_script.py
```

> The above will run the training script on two GPUs that live on a single machine and this is the barebones for performing distributed training with PyTorch.

Author: Went in the middle with the below clarification, phrasing as:
"The above will run the training script on two GPUs that live on a single machine and this is the barebones for performing only distributed training with PyTorch."

(Aka adding **only**)

Now let's talk about Accelerate, a library aimed to make this process more seameless and also help with a few best practices

## 🤗 Accelerate

> [Accelerate](https://huggingface.co/docs/accelerate) is a library designed to allow you to perform what we just did above, without needing to modify your code greatly. On top of this, the data pipeline innate to Accelerate can also improve performance to your code as well.

Reviewer: I believe mentioning the abstractions for dataloaders and metrics gathering is also important in addition to support for various device types and model related changes in distributed setup.

Author: Will do this in a followup as a deep dive, so that this can be surface level :)

    test_loader = torch.utils.data.DataLoader(test_dset, shuffle=False, batch_size=64)

    # Build model
    model = model.to(rank)
    ddp_model = DDP(model, device_ids=[rank])

    # Build optimizer
    optimizer = optim.AdamW(ddp_model.parameters(), lr=1e-3)

    # Train for a single epoch
    model.train()
    for batch_idx, (data, target) in enumerate(train_loader):
        data, target = data.to(device), target.to(device)
        output = model(data)
        loss = F.nll_loss(output, target)
        loss.backward()
        optimizer.step()
        optimizer.zero_grad()
    
    # Evaluate
    model.eval()
    correct = 0
    with torch.no_grad():
        for data, target in test_loader:
            data, target = data.to(device), target.to(device)
            output = model(data)
            pred = output.argmax(dim=1, keepdim=True)
            correct += pred.eq(target.view_as(pred)).sum().item()
    print(f'Accuracy: {100. * correct / len(test_loader.dataset)}')
```

Next let's talk about how Accelerate can help. There's a few issues with the above code:

1. This is slightly inefficient, given that `n` dataloaders are made based on each device and pushed.

> 2. This code will **only** work for multi-GPU, so special care would need to be made for it to be ran on a single node again, or on TPU.

Reviewer 2: Ah, it's mentioned here. Might be enough!

Accelerate helps this through the [`Accelerator`](https://huggingface.co/docs/accelerate/v0.12.0/en/package_reference/accelerator#accelerator) class. Through it, the code remains much the same except for three lines of code when comparing a single node to multinode, as shown below:

```python
def train_ddp_accelerate():

>     accelerator = Accelerator(device_placement=True)

Reviewer 2: ```suggestion
    accelerator = Accelerator()
```
This is the default.

Author: It is, but also in the docs where we assume no prior knowledge (in our docs) we have `device_placement=True`. 

Leave the default and make a note about device placement probably then?

Reviewer 2: Yeah, we can also update the docs since it was originally set like this to show how to remove the automatic device placement but no one ever complained about it.

    # Build DataLoaders
    transform = transforms.Compose([
        transforms.ToTensor(),
        transforms.Normalize((0.1307), (0.3081))
    ])

    train_dset = datasets.MNIST('data', train=True, download=True, transform=transform)
    test_dset = datasets.MNIST('data', train=False, transform=transform)

    train_loader = torch.utils.data.DataLoader(train_dset, shuffle=True, batch_size=64)
    test_loader = torch.utils.data.DataLoader(test_dset, shuffle=False, batch_size=64)

    # Build model
    model = BasicModel()

    # Build optimizer
    optimizer = optim.AdamW(model.parameters(), lr=1e-3)

    # Send everything through `accelerator.prepare`
    train_loader, test_loader, model, optimizer = accelerator.prepare(
        train_loader, test_loader, model, optimizer
    )

    # Train for a single epoch
    model.train()
    for batch_idx, (data, target) in enumerate(train_loader):
        output = model(data)
        loss = F.nll_loss(output, target)

>         loss.backward()

Reviewer: ```suggestion
        accelerator.backward()
```

        optimizer.step()
        optimizer.zero_grad()
    
    # Evaluate
    model.eval()
    correct = 0
    with torch.no_grad():
        for data, target in test_loader:

>             data, target = data.to(device), target.to(device)

Reviewer: ```suggestion
```

            output = model(data)
            pred = output.argmax(dim=1, keepdim=True)
            correct += pred.eq(target.view_as(pred)).sum().item()
    print(f'Accuracy: {100. * correct / len(test_loader.dataset)}')
```

Now any and all custom code needed to launch your code on any distributed setup is simplified to the `Accelerator` object. This code can then still be ran through the `torchrun` CLI, or through Accelerate's own CLI interface, [`accelerate launch`](https://huggingface.co/docs/accelerate/v0.12.0/en/basic_tutorials/launch).

> As a result its now trivialized to perform distributed training with Accelerate and keeping as much of the barebones PyTorch code the same as possible.

Reviewer: Needs rewording

Author: Fixed :)

> Earlier it was mentioned that Accelerate also makes the DataLoaders more efficient. This is through custom Samplers that will send parts of the batches automatically to different devices during training allowing for a single copy of the data to be known at one time, rather than four at once into memory.

Reviewer: I think here the important detail is that the user doesn't have to worry about making sure each device has a subset of the original dataset that is exclusive to it. Everything is handled under the hood. Also, single copy vs multiple copies isn't accurate because `DataLoaderDispatcher` (in case of IterableDataset) sends num_processes*batch_size data from rank 0 to all ranks leading to data deduplication/multiple copies wherein each rank then picks the slice exclusive to it.

To use it is as trivial as importing the launcher:

```python
from accelerate import notebook_launcher
```

And passing the training function we declared earlier, any arguments to be passed, and the number of processes to use (such as 8 on a TPU, or 2 for two GPUs). Both of the above training functions can be ran, but do note that after you start a single launch, the instance needs to be restarted before spawning another

```python
notebook_launcher(train_ddp, args=(), num_processes=2)
```

Or:

```python
notebook_launcher(train_accelerate_ddp, args=(), num_processes=2)
```

## Using 🤗 Trainer

Finally, we arrive at the highest level of API -- the Hugging Face [Trainer](https://huggingface.co/docs/transformers/main_classes/trainer).

This wraps as much training as possible while still being able to train on distributed systems without the user needing to do anything at all.

First we need to import the Trainer:

```python
from transformers import Trainer
```

> Then we define some TrainingArguments since it will handle the dataloader creation. The trainer also works through dictionaries, so a custom colalte function needs to be made.

Reviewer 2: ```suggestion
Then we define some `TrainingArguments` to control all the usual hyper-parameters. The trainer also works through dictionaries, so a custom collate function needs to be made.
```

    "Sets up the process group and configuration for PyTorch Distributed Data Parallelism"
    os.environ["MASTER_ADDR"] = 'localhost'
    os.environ["MASTER_PORT"] = "12355"

    # Initialize the process group
    dist.init_process_group("gloo", rank=rank, world_size=world_size)

def cleanup():
    "Cleans up the distributed environment"
    dist.destroy_process_group()
```

The last piece of the puzzle is *how do I send my data and model to another GPU?*

This is where the `DistributedDataParallel` module comes into play. It will copy your model onto each GPU, and when `loss.backward()` is called the backpropagation is performed and the resulting gradients across all these copies of the model will be averaged/reduced. This ensures each device has the same weights post the optimizer step.

Below is an example of our training setup, refactored as a function, with this capability:

> Note: Here rank is the overall rank of the current GPU compared to all the other GPUs available, meaning they have a rank of `0 -> n-1`

```python
from torch.nn.parallel import DistributedDataParallel as DDP

def train(model, rank, world_size):
    setup(rank, world_size)
    model = model.to(rank)

>     ddp_model = DDP(model, device_ids=[rank])

Reviewer 3: Thanks @muellerzr for putting up the blog post!!

I just wanna mention a few more techniques to further speed up DDP. No need to change version if the goal is not squeezing out the last bit of perf from the cluster. 

1. Setting `gradient_as_bucket_view` to `True`. This can make `param.grad` points to an offset of DDP's bucket, which can help save two copies (i.e., grad to bucket, bucket to grad).
2.  It is possible to overlap optimizer step with DDP's backward communication by using the DDP comm hook. (https://pytorch.org/docs/stable/generated/torch.nn.parallel.DistributedDataParallel.html#torch.nn.parallel.DistributedDataParallel.register_comm_hook) However, this will only make a difference if communication delay is dominating.

Author: Thanks @mrshenli! The goal of this one is not, but I'll definitely include it in the next! (And look at having this happen also in `accelerate`), thanks for the TIL! 🤗
